pagopa/pn-frontend
# PN-2261 Handling failure in API calls to retrieve information, for pa-webapp

## Short description
Use in pa-webapp the tools to handle the cases in which an API call made to fetch information fails, i.e. yields a status code in the 4xx or 5xx ranges.
In most cases, some part of the page is replaced by a banner which indicates the inability of fetching data, and allows the user to retry the data fetch. In the particular case of the call which fetches the ToS accept status, a failure implies an immediate logout.

## List of changes proposed in this pull request
- Apply the `<ApiErrorWrapper>` or `<ApiError>` components, in the latter case with ad-hoc handling of the API call error (obtained through the `useErrors` custom hook, for the API calls regarding: fetch of a list of notifications, fetch of the details of one modification, fetch of the user groups used to create a notification, fetch of the organization party corresponding to the logged user. In the latter case, the `ApiErrorWrapper` is used inside the router, an ad-hoc guard was implemented to this end.
- Use the `performThunkAction` to simplify the development of some Redux thunk actions, and to avoid forgetting to soundly manage the eventual exceptions.
- Added a `renderWithoutRouter` in `src/__test__/test-utils.tsx`, it simplifies the writing of tests for which we want to set an initial Redux store state, but we want to use a react-router Router ad-hoc for the test, and not the `BrowserRouter` defined in the `render` function inside the same file.

## How to test
You should be able to simulate a failure in one API call. The involved calls are:
- `getOrganizationParty` (in ExternalRegistries), this is invoked when entering the app.
- `getSentNotifications`, to obtain the list of notifications.
- `getSentNotification`, to obtain the detail of one notification.
- `getUserGroups`, to obtain the options for the group select in the workflow of manual creation of a notification, in the first step.
- `getConsentByType` (in Consents), to obtain the ToS acceptance status.

The next-to-last commit includes mock versions for several of these API calls.
